File: which.py
import fnmatch
import os
import sys



def usage() :
    print 'which -p <command>'
    exit()


def check_exact(dir, file2find) :
    if show_dirs :
        print 'checking in ' + dir
    try:
        for filename in os.listdir(dir):
            if filename == file2find:
                return True
        return False
    except Exception as e:
        print e
        return False


def find_in_dirs(dirs, file2find) :
    for dir in dirs :
        result = check_exact(dir, file2find)
        if result == True :
            print 'found in ' + dir + '\\' + file2find
            return True
    print file2find + ' not found.'
    return False




roots = [] # ["c:\\bin", "c:\\bin\py"]
show_dirs = False
file2find = ''

if len(sys.argv) == 3:
    if sys.argv[1] == '-p':
        show_dirs = True
        file2find = sys.argv[2]
    else :
        usage()
else :
    file2find = sys.argv[1]
    
path_dirs = (os.environ['PATH']).split(';')
for dir in path_dirs:
    if dir != '':
        roots.append(dir)

find_in_dirs(roots, file2find)

File: which.js
var fs = require('fs')

function log(msg) { console.log(msg) }

function usage(msg) {
    if (msg)
        log(msg)
    log("Usage: " + __filename + " path/to/directory")
    process.exit(-1)
}

function check_exact(dir, cmd) {
    try {
        let items = fs.readdirSync(dir)
        return (items.filter(i => i == cmd).length == 1)
    } catch (ex) {
        log('error: ' + ex)
    }

    return false
}

let show_paths = false
let cmd = "";
let syspath = process.env.path
let syspath_folders = syspath.split(';').filter(dir => dir != "")
let found = false

if (process.argv.length <= 2)
    usage("")

for (let i = 2; i < process.argv.length; i++) {
    if (process.argv[i] == '-p')
        show_paths = true
    else {
        if (cmd != "")
            usage("command set twice")
        cmd = process.argv[i]
    }
}

for (let i = 0; i < syspath_folders.length; i++) {
    let dir = syspath_folders[i]

    if (show_paths)
        log(dir)

    if (check_exact(dir, cmd)) {
        log('Found in: ' + dir)
        found = true
        break
    }
}

if (!found)
    log('command not found')
